pkg/sfu/buffer/videolayerutils.go

```diff
@@ -0,0 +1,306 @@
+package buffer
+
+import (
+	"github.com/livekit/protocol/livekit"
+	"github.com/livekit/protocol/logger"
+)
+
+const (
+	QuarterResolution = "q"
+	HalfResolution    = "h"
+	FullResolution    = "f"
+)
+
+func LayerPresenceFromTrackInfo(trackInfo *livekit.TrackInfo) map[livekit.VideoQuality]bool {
+	if trackInfo == nil || len(trackInfo.Layers) == 0 {
+		return nil
+	}
+
+	layerPresence := make(map[livekit.VideoQuality]bool, livekit.VideoQuality_HIGH+1)
+	for _, layer := range trackInfo.Layers {
+		layerPresence[layer.Quality] = true
+	}
+
+	return layerPresence
+}
+
+func RidToSpatialLayer(rid string, trackInfo *livekit.TrackInfo) int32 {
+	lp := LayerPresenceFromTrackInfo(trackInfo)
+	if lp == nil {
+		switch rid {
+		case QuarterResolution:
+			return 0
+		case HalfResolution:
+			return 1
+		case FullResolution:
+			return 2
+		default:
+			return 0
+		}
+	}
+
+	switch rid {
+	case QuarterResolution:
+		switch {
+		case lp[livekit.VideoQuality_LOW] && lp[livekit.VideoQuality_MEDIUM] && lp[livekit.VideoQuality_HIGH]:
+			fallthrough
+		case lp[livekit.VideoQuality_LOW] && lp[livekit.VideoQuality_MEDIUM]:
+			fallthrough
+		case lp[livekit.VideoQuality_LOW] && lp[livekit.VideoQuality_HIGH]:
+			fallthrough
+		case lp[livekit.VideoQuality_MEDIUM] && lp[livekit.VideoQuality_HIGH]:
+			return 0
```

Comment:
This can be written more succinctly, but I chose to list it out explicitly to make it very clear what cases are handled and how.